include/cuco/detail/open_addressing/functors.cuh

```diff
@@ -44,9 +47,64 @@ struct get_slot {
   {
     auto const window_idx = idx / StorageRef::window_size;
     auto const intra_idx  = idx % StorageRef::window_size;
-    return storage_[window_idx][intra_idx];
+    if constexpr (HasPayload) {
+      auto const& [first, second] = storage_[window_idx][intra_idx];
+      return thrust::make_tuple(first, second);
+    } else {
+      return storage_[window_idx][intra_idx];
+    }
+  }
+};
+
+/**
+ * @brief Device functor returning whether the given slot is filled
+ *
+ * @tparam HasPayload Flag indicating whether the slot contains a payload
+ * @tparam T The slot key type
+ */
+template <bool HasPayload, typename T>
+struct slot_is_filled {
+  T empty_sentinel_;   ///< The value of the empty key sentinel
+  T erased_sentinel_;  ///< Key value that represents an erased slot
+
+  /**
+   * @brief Constructs `slot_is_filled` functor with the given sentinels
+   *
+   * @param empty_sentinel Key sentinel indicating an empty slot
+   * @param erased_sentinel Key sentinel indicating an erased slot
+   */
+  explicit constexpr slot_is_filled(T const& empty_sentinel, T const& erased_sentinel) noexcept
+    : empty_sentinel_{empty_sentinel}, erased_sentinel_{erased_sentinel}
+  {
+  }
+
+  /**
+   * @brief Indicates if the target slot `slot` is filled.
+   *
+   * @tparam S The slot type
+   *
+   * @param slot The slot
+   *
+   * @return `true` if slot is filled
+   */
+  template <typename S>
+  __device__ constexpr bool operator()(S const& slot) const noexcept
+  {
+    auto const key = [&]() {
+      if constexpr (HasPayload) {
+        // required by thrust zip iterator in `retrieve_all`
+        if constexpr (cuco::detail::is_thrust_pair_like<S>::value) {
+          return thrust::get<0>(slot);
+        } else {
+          return slot.first;
+        }
+      } else {
+        return slot;
+      }
+    }();
```

Comment:
Do we need to guard the lambda with `cuda::proclaim_return_type`? @bdice probably knows the answer the best.

Reply:
Now that we're on CCCL 2.2.0 as a minimum requirement, you will see build errors if you need `cuda::proclaim_return_type` and don't have it. In other words, if this is building, you're fine!

Reply:
@bdice Realized it's the case after asking you. Thanks for the confirmation!